File: EMVS/DataEMVSS.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#define HPC 0
#define LPC 1
#define Emory 2

int main()
{
	char fname[100];
	char Rname[100];
	char line[1024];
	char master[100];
	char method[20];
	char acronym[20];
	char home[100];
	char script[100];
	char datadir[100];
	char data[20];
	char src[50];
	char vname[50];
	FILE *f, *g;
	int i, K, where;
	double eta;

	K = 5;
	eta = 0.0;

	strcpy(data,"TCGA");
	strcpy(method,"EMVSS");
	sprintf(acronym,"%s_%s_%.1f_CV%d",method,data,eta,K);

	if ( access("/home/cchan40",X_OK) == 0 )
	{
		where = Emory;
		strcpy(master,"/home/cchan40/project/EMSHS");
	}
	else if ( access("/project/qlonglab/changgee",X_OK) == 0 )
	{
		where = HPC;
		strcpy(master,"/home/changgee/project/EMSHS");
	}
	else
	{
		where = LPC;
		strcpy(master,"/home/changgee/project/EMSHS");
	}
	sprintf(home,"%s/EMVS",master);
	sprintf(datadir,"%s/datasets/%s",master,data);
	sprintf(script,"%s/%s",home,data);
	strcpy(src,"DataEMVSS.R");
	sprintf(vname,"res%s",acronym);

	sprintf(fname,"%s",acronym);
	g = fopen(fname,"w");
	chmod(fname,0755);

	for ( i=0 ; i<K ; i++ )
	{
		sprintf(fname,"%s/%s_%02d",script,acronym,i+1);
		if ( where == Emory )
			sprintf(line,"qsub -q fruit.q %s\n",fname);
		else if ( where == HPC )
			sprintf(line,"bsub -q qlonglab -e %s.e -o %s.o < %s\n",fname,fname,fname);
		else
			sprintf(line,"bsub -q cceb_normal -e %s.e -o %s.o < %s\n",fname,fname,fname);
		fputs(line,g);

		f = fopen(fname,"w");
		if ( where == LPC )
			fputs("module load R\n",f);
		else if ( where == HPC )
		{
			fputs("source /etc/profile.d/modules.sh\n",f);
			fputs("module load R-3.3.1\n",f);
		}
		sprintf(Rname,"%s.R",fname);
		sprintf(line,"R --vanilla < %s\n",Rname);
		fputs(line,f);
		fclose(f);

		f = fopen(Rname,"w");
		fputs("library(compiler)\n",f);
		fputs("enableJIT(3)\n",f);
		sprintf(line,"source(\"%s/%s\")\n",home,src);
		fputs(line,f);
		sprintf(line,"load(\"%s/data\")\n",datadir);
		fputs(line,f);
		sprintf(line,"load(\"%s/fold%d\")\n",datadir,K);
		fputs(line,f);

		fputs("v0 = exp(seq(log(0.001),log(100),length.out=20))\n",f);
		fputs("v1 = 1000\n",f);
		sprintf(line,"eta = %.1f\n",eta);
		fputs(line,f);
		sprintf(line,"%s = DataEMVSS(y,X,v0,v1,eta,E,fold=f,k=%d)\n",vname,i+1);
		fputs(line,f);

		sprintf(line,"save(%s,file=\"%s/%s%02d\")\n",vname,script,vname,i+1);
		fputs(line,f);
		fclose(f);
	}
	fclose(g);

	sprintf(fname,"%sMERGE.R",acronym);
	g = fopen(fname,"w");

	fputs("library(abind)\n",g);

	sprintf(line,"for ( i in 1:%d )\n",K);
	fputs(line,g);
	fputs("{\n",g);
	sprintf(line,"  fname = sprintf(\"%s/%s%%02d\",i)\n",script,vname);
	fputs(line,g);
	fputs("  load(fname)\n",g);
	fputs("  if ( i==1 )\n",g);
	fputs("  {\n",g);
	sprintf(line,"    tmp = %s\n",vname);
	fputs(line,g);
	fputs("  }\n",g);
	fputs("  else\n",g);
	fputs("  {\n",g);
	sprintf(line,"    tmp$k = c(tmp$k,%s$k)\n",vname);
	fputs(line,g);
	sprintf(line,"    tmp$beta = abind(tmp$beta,%s$beta,along=2)\n",vname);
	fputs(line,g);
	sprintf(line,"    tmp$L = abind(tmp$L,%s$L,along=1)\n",vname);
	fputs(line,g);
	sprintf(line,"    tmp$SSPECV = tmp$SSPECV + %s$SSPECV\n",vname);
	fputs(line,g);
	sprintf(line,"    tmp$time = tmp$time + %s$time\n",vname);
	fputs(line,g);
	fputs("  }\n",g);
	fputs("}\n",g);
	sprintf(line,"%s = tmp\n",vname);
	fputs(line,g);
	sprintf(line,"save(%s,file=\"%s/%s\")\n",vname,home,vname);
	fputs(line,g);

	fclose(g);

	sprintf(fname,"%sMERGE",acronym);
	f = fopen(fname,"w");
	fputs("module load R\n",f);
	sprintf(line,"R --vanilla < %s/%sMERGE.R\n",home,acronym);
	fputs(line,f);
	fclose(f);
}

File: EMVS/DataEMVSS.R
  if ( file.exists("~/project/EMSHS/EMVS/EMVS_Rcode_original.R") )
  source("~/project/EMSHS/EMVS/EMVS_Rcode_original.R")
if ( file.exists("EMVS/EMVS_Rcode_original.R") )
  source("EMVS/EMVS_Rcode_original.R")

DataEMVSS <- function(y,X,w,v0,v1,eta=0,E=NULL,fold,k)
{
  r = ncol(fold)
  K = length(k)
  n = nrow(X)
  p = ncol(X)
  
  D1 = length(v0)
  D2 = length(eta)
  SSPECV = array(0,c(D1,D2,r))
  beta = array(0,c(p,K,D1,D2,r))
  L = array(0,c(K,D1,D2,r))  

  before = proc.time()
  
  if ( !is.null(E))
  {
    G = matrix(0,p,p)
    G[E[,1]+(E[,2]-1)*p] = 1
  }
  
  for ( i in 1:r )
  {
    for ( j in 1:K )
    {
      ik = which(fold[,i]==k[j])
      
      for ( d2 in 1:D2 )
      {
        if ( !is.null(E) & eta[d2] > 0 )
          fit <- EMVS(y[-ik]*w[-ik],X[-ik,]*w[-ik],v0=v0,v1=v1,type="MRF",mu=0,Sigma=G*eta[d2],sigma_init=1,epsilon=1e-4,v1_g=v1)
        else
          fit <- EMVS(y[-ik]*w[-ik],X[-ik,]*w[-ik],v0=v0,v1=v1,type="betabinomial",sigma_init=1,epsilon=1e-4,a=1,b=1)
      
        beta[,j,,d2,i] = t(fit$betas*(fit$p>0.5))
        L[j,,d2,i] = apply(fit$p>0.5,1,sum)
        yhat = X[ik,] %*% beta[,j,,d2,i]
        SSPECV[,d2,i] = SSPECV[,d2,i] + apply((y[ik]-yhat)^2,2,sum)
      }
    }
  }
  
  after = proc.time()-before
  
  list(r=r,p=p,n=n,K=K,fold=fold,k=k,v0=v0,v1=v1,eta=eta,beta=beta,L=L,SSPECV=SSPECV,time=after)
}

